From 716bf55d678ae0ecef906d6b37e9f05e035787e2 Mon Sep 17 00:00:00 2001
From: Jan-Tarek Butt <tarek@ring0.de>
Date: Mon, 16 Jul 2018 02:38:45 +0200
Subject: [PATCH] add package gluon-geolocator:

commit 9009a58268c48a2317ac8fa0814ebfce2236e1c6
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Mon Jul 16 02:30:10 2018 +0200

    gluon-geolocator: request wifi interface once per radio and via ubus

commit 9980ed03698e08226fad6107123780c4cf433d56
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Mon Jul 16 01:33:28 2018 +0200

    gluon-geolocator: update to master package configuration

    * drop bssid blacklist
    * only allow Master mode scanned wifis for osition request.

commit 33fcb193d58e02db1688d45950e3fa1c88a01833
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Mon Feb 19 18:03:12 2018 +0100

    geolocator: rebase site_check

commit 8ef746bbd9d75bed48592e5442b2ce1324f6c5a4
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Wed Feb 14 07:18:19 2018 +0100

    geolocator: set default values in config

commit ee62a2f23a38b03d6d344b0b83e117d59ba4e02b
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Tue Feb 13 08:59:19 2018 +0100

    geolocator: set table inserts in brackes

commit e8dabe512c4a461edff364f103088f5a5ae877da
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Wed Jan 24 12:15:39 2018 +0100

    geolocator: ensure MAC addreses are filert by uppercase matching

commit 6bf2536776f881a4cb527497187644b81391ae6b
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Tue Dec 26 23:05:42 2017 +0100

    gluon-geolocator: fix intervall in 540-geolocator-settings

     * change intervall value from seconds to minutes

commit dc71246534bba2a561ec4cbabe915d4504fd5a76
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Tue Dec 26 22:51:26 2017 +0100

    gluon-geolocator: refactoring and spellcheck

     * rm irrelevant else states
     * fix print message for time stamp file

commit af55a3609768b4563039fbb637bac20525c84e21
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sun Sep 3 14:45:17 2017 +0200

    geolocator: rm TODO commend after discussion wiht Neoraider

commit 6bb40bc4ec73f4162e9e6da5b639811d85b072e7
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Aug 18 15:44:59 2017 +0200

    gluon-geolocator: set uci getbool fix scan for surrounded wifis

commit 1a4b32d669e373e92eaefd3eeb1b29e3cc1c4a48
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Thu Aug 17 09:12:37 2017 +0200

    gluon-geolocator: fix line endings and write mode

commit a55e61b17bd2e63b54bb6e964420659adb7c84eb
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Wed Aug 16 23:54:46 2017 +0200

    gluon-geolocator: rm .sh file ending

commit 1a68d992e77b56ef36d53f3f88a0921d5cd9d922
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Wed Aug 16 23:54:08 2017 +0200

    gluon-geolocator: porting from shell to lua code

commit f2d8b1bfab984b746bf61f4fdaaebf2a110f3eb8
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Wed Aug 16 23:53:29 2017 +0200

    gluon-geolocator: rm geolocator.sh

commit 462f11fd9b6223c86413300aa8ae90cc3a020885
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sun Aug 13 17:11:32 2017 +0200

    gluon-geolocator: upgrade script prepare for new site file

commit 46817830cb12e2d8754e34e2578b3aabf4d5042b
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sat Aug 12 21:23:49 2017 +0200

    gluon-geolocator: geolocator.sh use if/fi instead of {}

commit ed586f025878013f36acfdd7b1b2ba59d2715059
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sat Aug 12 21:22:04 2017 +0200

    gluon-geolocator: geolocator.sh rm unneeded ;

commit fb6cc92ead8fa0efb750685074998fcbe0cd3dc6
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sat Aug 12 21:20:59 2017 +0200

    gluon-geolocator: geolocator.sh replace spaces with taps

commit 205aff7f9e30aa722cbd134df1fedc19083bf18e
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sat Aug 12 20:55:00 2017 +0200

    gluon-geolocator: use taps and chang val type in check_site.lua

     * set geolocator.autolocation as need_boolean
     * replace spaces with taps

commit 18cf80032b2d770adba545c7c0e92ccb2d66c306
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Sat Aug 12 20:51:13 2017 +0200

    gluon-geolocator: clean up cron file

     * rm unneded semicolon
     * remove the sh

commit 13c359c576ef6f27190f0eacc2140303b0f17156
Author: Tata <tarek@ring0.de>
Date:   Sat Aug 12 18:37:24 2017 +0200

    gluon-geolocator: change lua require

    540-geolocator-settings: change lua require 'gluon.site_config' to 'gluon.site'

commit 91afad0f11daf091526c7b6b7e28b991516a2829
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Aug 4 20:09:29 2017 +0200

    gluon-geolocator: rm ffnw spec stuff and add site array to geolocator.sh

commit 3bb37e65e6098a3f20a8ad079f109bdd2483528f
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Aug 4 20:08:32 2017 +0200

    gluon-geolocator: change check_site.lua

commit 8a49856bd80cac30aeeed32e1f45bbb74029dbd1
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Aug 4 20:08:08 2017 +0200

    gluon-geolocator: add luaminifyer

commit acb9d3334fbec0ea67190239c112d540a5111815
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Aug 4 20:07:36 2017 +0200

    add 540-geolocator-settings

commit 2aff2f0e7653a9497f3f89e68e3b8b49b771c230
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Aug 4 20:07:05 2017 +0200

    mv to 540-geolocator-settings

commit 4be5d26209e5788af6ae1811b340ba75670e62c8
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Jul 28 02:45:02 2017 +0200

    gluon-geolocator: add upgrade script for intervall

commit ec078c88bc285693875fea8a50e2cef98238f2bb
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Jul 28 02:44:32 2017 +0200

    gluon-geolocator: add cron file

commit a011b86a3013a5513866a083afb8907a95f15281
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Jul 28 02:43:16 2017 +0200

    gluon-geolocator: add geolocator.sh

commit a0abc6d6684d78492f7a4594a571d843beae8db9
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Jul 28 02:42:47 2017 +0200

    gluon-geolocator: add config geolocator

commit 7584e637eb0f9987bc856b82545b71a6a9c869dd
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Jul 28 02:41:58 2017 +0200

    gluon-geolocator: add check_site.lua

commit 7931825cae52897d926d7357417fd57f2f9802f2
Author: Jan-Tarek Butt <tarek@ring0.de>
Date:   Fri Jul 28 02:41:30 2017 +0200

    add new pkg gluon-geolocator

Signed-off-by: Jan-Tarek Butt <tarek@ring0.de>
---
 package/gluon-geolocator/Makefile             |  14 ++
 package/gluon-geolocator/check_site.lua       |   2 +
 .../files/etc/config/geolocator               |   4 +
 .../files/usr/lib/micron.d/geolocator         |   1 +
 .../luasrc/lib/gluon/geolocator/geolocator    | 127 ++++++++++++++++++
 .../lib/gluon/upgrade/540-geolocator-settings |  26 ++++
 6 files changed, 174 insertions(+)
 create mode 100644 package/gluon-geolocator/Makefile
 create mode 100644 package/gluon-geolocator/check_site.lua
 create mode 100644 package/gluon-geolocator/files/etc/config/geolocator
 create mode 100644 package/gluon-geolocator/files/usr/lib/micron.d/geolocator
 create mode 100755 package/gluon-geolocator/luasrc/lib/gluon/geolocator/geolocator
 create mode 100755 package/gluon-geolocator/luasrc/lib/gluon/upgrade/540-geolocator-settings

diff --git a/package/gluon-geolocator/Makefile b/package/gluon-geolocator/Makefile
new file mode 100644
index 00000000..2f6e48d7
--- /dev/null
+++ b/package/gluon-geolocator/Makefile
@@ -0,0 +1,14 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=gluon-geolocator
+PKG_VERSION:=1
+PKG_RELEASE:=1
+
+include ../gluon.mk
+
+define Package/gluon-geolocator
+  TITLE:=Provide the geolocator to receive positions over wifi
+  DEPENDS:=+gluon-node-info +micrond
+endef
+
+$(eval $(call BuildPackage,gluon-geolocator))
diff --git a/package/gluon-geolocator/check_site.lua b/package/gluon-geolocator/check_site.lua
new file mode 100644
index 00000000..2771d938
--- /dev/null
+++ b/package/gluon-geolocator/check_site.lua
@@ -0,0 +1,2 @@
+need_boolean(in_site({'geolocator', 'autolocation'}), false)
+need_number(in_site({'geolocator', 'interval'}), false)
diff --git a/package/gluon-geolocator/files/etc/config/geolocator b/package/gluon-geolocator/files/etc/config/geolocator
new file mode 100644
index 00000000..dc7ae510
--- /dev/null
+++ b/package/gluon-geolocator/files/etc/config/geolocator
@@ -0,0 +1,4 @@
+config geolocator 'settings'
+	option static_location '0'
+	option auto_location '1'
+	option refresh_interval '1000'
diff --git a/package/gluon-geolocator/files/usr/lib/micron.d/geolocator b/package/gluon-geolocator/files/usr/lib/micron.d/geolocator
new file mode 100644
index 00000000..7379b730
--- /dev/null
+++ b/package/gluon-geolocator/files/usr/lib/micron.d/geolocator
@@ -0,0 +1 @@
+*/5 * * * *     /lib/gluon/geolocator/geolocator
diff --git a/package/gluon-geolocator/luasrc/lib/gluon/geolocator/geolocator b/package/gluon-geolocator/luasrc/lib/gluon/geolocator/geolocator
new file mode 100755
index 00000000..7b15def7
--- /dev/null
+++ b/package/gluon-geolocator/luasrc/lib/gluon/geolocator/geolocator
@@ -0,0 +1,127 @@
+#!/usr/bin/lua
+
+local uci = require('simple-uci').cursor()
+local json = require ("luci.jsonc")
+local ubus = require 'ubus'
+local iwinfo = require("iwinfo")
+
+local LOC="gluon-node-info"
+local GLC="geolocator"
+
+if not uci:get_bool(GLC, "settings", "auto_location") then
+  os.exit(0)
+end
+
+-- PID file to ensure the geolocator isn't running parallel
+local PID_PART="/var/run/geolocator.pid"
+local TIME_STAMP="/tmp/geolocator_timestamp"
+
+local function file_exsist(file)
+  return io.open(file, "r") ~= nil
+end
+
+if file_exsist(PID_PART) then
+  io.stdout:write("The geolocator is still running.\n")
+  os.exit(0)
+end
+if io.open(PID_PART, "w") == nil then
+  io.stdout:write("Can`t create pid file on " .. PID_PART .. "\n")
+  os.exit(1)
+end
+
+-- Program terminating function including removing of PID file
+local function exit(exc)
+  if file_exsist(PID_PART) then
+    os.remove(PID_PART)
+  end
+  os.exit(exc)
+end
+
+-- Iterates over all active WLAN interfaces
+-- Returning true from the callback function will skip all remaining
+-- interfaces of the same radio
+local function foreach_radio(f)
+  local uconn = assert(ubus.connect(), 'failed to connect to ubus')
+  local status = uconn:call('network.wireless', 'status', {})
+  ubus.close(uconn)
+
+  for _, radio in pairs(status) do
+    for _, iface in ipairs(radio.interfaces) do
+      if f(iface.ifname) then
+        break
+      end
+    end
+  end
+end
+
+local function receive_json(request)
+  local f = assert(io.popen(string.format("exec wget -T 15 -q -O- '%s'", request)), 'failed to run wget')
+  local data = f:read('*a')
+  f:close()
+
+  return json.parse(data)
+end
+
+-- Get position
+local function locate()
+  local done_bssids = {}
+  local found_bssids = {}
+  foreach_radio(function(ifname)
+    local iw = iwinfo[iwinfo.type(ifname)]
+    if not iw then
+      -- Skip other ifaces of this radio, as they
+      -- will have the same type
+      return true
+    end
+
+    local scanlist = iw.scanlist(ifname)
+    if not scanlist then
+      return false
+    end
+
+    for _, entry in ipairs(scanlist) do
+      if entry.mode:match("Master") then
+        local bssid = string.upper(entry.bssid:gsub(":", ""))
+        if not done_bssids[bssid] then
+          table.insert(found_bssids, bssid)
+          done_bssids[bssid] = true
+        end
+      end
+    end
+
+    return true
+  end)
+
+  assert(#found_bssids >= 12, 'insufficient BSSIDs found')
+
+  local data = receive_json('http://openwifi.su/api/v1/bssids/' .. table.concat(found_bssids, ','))
+  assert(type(data) == 'table' and data.lon and data.lat, 'location not available')
+
+  return data
+end
+
+-- Check if interval over or not exist
+if file_exsist(TIME_STAMP) then
+  if os.time() - tonumber(io.open(TIME_STAMP):read("*a")) < uci:get(GLC, "settings", "refresh_interval") * 60 then
+    exit(0)
+  end
+end
+
+local pos = locate()
+if not next(pos) then
+  exit(1)
+end
+if not uci:get_bool(GLC, "settings", "static_location") then
+  uci:set(LOC, uci:get_first(LOC, 'location'), 'latitude', pos.lat)
+  uci:set(LOC, uci:get_first(LOC, 'location'), 'longitude', pos.lon)
+  uci:save(LOC)
+  uci:commit(LOC)
+end
+local timestap = io.open(TIME_STAMP, "w")
+if timestap ~= nil then
+  timestap:write(os.time())
+  timestap:close()
+  exit(0)
+end
+io.stdout:write("Can`t create file on " .. TIME_STAMP .. "\n")
+exit(1)
diff --git a/package/gluon-geolocator/luasrc/lib/gluon/upgrade/540-geolocator-settings b/package/gluon-geolocator/luasrc/lib/gluon/upgrade/540-geolocator-settings
new file mode 100755
index 00000000..84c9f99c
--- /dev/null
+++ b/package/gluon-geolocator/luasrc/lib/gluon/upgrade/540-geolocator-settings
@@ -0,0 +1,26 @@
+#!/usr/bin/lua
+
+local site = require 'gluon.site'
+local uci = require('simple-uci').cursor()
+
+local config = 'geolocator'
+
+local static_location = uci:get(config, 'settings', 'static_location')
+
+local auto_location = uci:get(config, 'settings', 'auto_location')
+if not auto_location then
+	auto_location = site.geolocator.autolocation(false)
+end
+
+local refresh_interval = uci:get(config, 'settings', 'refresh_interval')
+if not refresh_interval then
+	refresh_interval = site.geolocator.interval(720) -- default: 12h
+end
+
+uci:delete(config, 'settings')
+uci:section(config, config, 'settings', {
+	static_location = static_location,
+	refresh_interval = refresh_interval,
+	auto_location = auto_location,
+})
+uci:save(config)
-- 
2.18.0

